File: stack/0155_min_stack/0155_min_stack.go
package main

import "fmt"

// MinStack define
type MinStack struct {
	elements []int
	min      []int
	l        int
}

// Contructor define
func Contructor() MinStack {
	return MinStack{make([]int, 0), make([]int, 0), 0}
}

// Push define
func (this *MinStack) Push(x int) {
	this.elements = append(this.elements, x)
	if this.l == 0 {
		this.min = append(this.min, x)
	} else {
		min := this.GetMin()
		if x < min {
			this.min = append(this.min, x)
		} else {
			this.min = append(this.min, min)
		}
	}
	this.l++
}

// Pop define
func (this *MinStack) Pop() {
	this.l--
	this.min = this.min[:this.l]
	this.elements = this.elements[:this.l]
}

func (this *MinStack) Top() int {
	return this.elements[this.l-1]
}

func (this *MinStack) GetMin() int {
	return this.min[this.l-1]
}

func main() {
	minStack := Contructor()
	minStack.Push(1)
	minStack.Push(2)
	minStack.Push(3)
	fmt.Println(minStack.Top())
	fmt.Println(minStack.GetMin())
	minStack.Pop()
	fmt.Println(minStack.Top())
	fmt.Println(minStack.GetMin())
}

File: stack/0155_min_stack/0155_min_stack.py
class MinStack(object):
    def __init__(self, s=[]):
        self.s = s
        self.t = []
        self.m = float('inf')

    def push(self, x):
        self.s.append(x)
        if x < self.m:
            self.m = x
        self.t.append(self.m)

    def pop(self):
        self.s.pop()
        self.t.pop()
        self.m = self.t[-1] if self.t else float('inf')

    def top(self):
        return self.s[-1] if self.s else None

    def getMin(self):
        return self.m if self.s else None


minStack = MinStack()
minStack.push(1)
minStack.push(2)
minStack.push(3)
print(minStack.top())
print(minStack.getMin())
minStack.pop()
print(minStack.top())
print(minStack.getMin())